+nosnoc/+discrete_problem/stewart.m

```diff
@@ -0,0 +1,771 @@
+                obj.f = obj.f + t_stage*dcs.f_lsq_x_fun(obj.w.x(ii,opts.N_finite_elements(ii),opts.n_s),...
+                    model.x_ref_val(:,ii),...
+                    p);
+                obj.f = obj.f + t_stage*dcs.f_lsq_u_fun(obj.w.u(ii),...
+                    model.u_ref_val(:,ii),...
+                    p);
+                if ~opts.cost_integration
+                    obj.f = obj.f + dcs.f_q_fun(x_ijk, z_ijk, lambda_ijk, theta_ijk, mu_ijk, ui, v_global, p);
+                end
+
+                % Clock <Constraints
+                % TODO(@anton) HERE BE DRAGONS. This is by far the worst part of current nosnoc as it requires the discrete problem
+                %              to understand something about the time-freezing reformulation which is ugly.
+                if obj.opts.use_fesd && opts.equidistant_control_grid
+                    if opts.time_optimal_problem
+                        if opts.use_speed_of_time_variables
+                            obj.g.equidistant_control_grid(ii) = {[sum_h - opts.h;sot*sum_h - obj.w.T_final()/opts.N_stages]};
+                        else
+                            obj.g.equidistant_control_grid(ii) = {sum_h - obj.w.T_final()/opts.N_stages};
+                        end
+                    else
+                        if opts.relax_terminal_numerical_time
+                            % TODO(@armin) why is this allowed to be negative?
+                            obj.w.s_numerical_time(ii) = {{'s_numerical', 1}, -2*opts.h, 2*opts.h, opts.h/2};
+                            g_eq_grid = [sum_h - t_stage - obj.w.s_numerical_time(ii);
+                                -(sum_h - t_stage) - obj.w.s_numerical_time(ii)];
+                            obj.g.equidistant_control_grid(ii) = {g_eq_grid, -inf, 0};
+                            obj.f = obj.f + opts.rho_terminal_numerical_time*obj.w.s_numerical_time(ii);
+                        else
+                            obj.g.equidistant_control_grid(ii) = {t_stage-sum_h};
+                        end
+                    end
+                end
+            end
+
+            % Terminal cost
+            obj.f = obj.f + dcs.f_q_T_fun(obj.w.x(ii,jj,kk), obj.w.z(ii,jj,opts.n_s), v_global, p_global);
+
+            % Terminal_lsq_cost
+            obj.f = obj.f + h0*opts.N_finite_elements(ii)*dcs.f_lsq_T_fun(obj.w.x(ii,jj,opts.n_s),...
+                model.x_ref_end_val,...
+                p);
+
+            % Terminal constraint
+            if opts.relax_terminal_constraint_homotopy
+                error("Currently unsupported")
+            end
+            g_terminal = dcs.g_terminal_fun(obj.w.x(opts.N_stages,opts.N_finite_elements(opts.N_stages),opts.n_s+rbp),...
+                obj.w.z(opts.N_stages,opts.N_finite_elements(opts.N_stages),opts.n_s+rbp),...
+                v_global, p_global);
+            switch opts.relax_terminal_constraint
+              case 0 % hard constraint
```

Comment:
we should use an ENUM insetaed of 0,1,2,3 (the same key words can be then used when we have latter the general constraint relaxations), set e.g. via different options.

opts.terminal_constraint_relaxation_type = 'None',  'Ell_1', 'Ell_2', 'Ell_inf',  'None' is default